### PR TITLE
Make format_args!("literal") const.

### PR DESCRIPTION
**Pull request by m-ou-se**
This makes `format_args!("literal")` const, while leaving all other invocations of format_args!() non-const.

cc @rust-lang/wg-const-eval

### CODE REVIEW

File: src/test/ui/consts/const-eval/format_args.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#![crate_type = "lib"]
+
+const A: std::fmt::Arguments = format_args!("literal");
+
+const B: std::fmt::Arguments = format_args!("{}", 123);
+//~^ ERROR calls in constants are limited to
+//~| ERROR calls in constants are limited to
+//~| ERROR temporary value
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/format_args.rs:8]: Why are there 3 errors here and not just the one we'd expect? I guess that comes from the extra code for the argument `123`? "temporary value dropped while borrowed" is a bit odd though...

File: src/test/ui/consts/const-eval/format_args.rs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#![crate_type = "lib"]
+
+const A: std::fmt::Arguments = format_args!("literal");
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/format_args.rs:3]: So this is insta-stable?

Comment (oli-obk) [reply src/test/ui/consts/const-eval/format_args.rs:3]: i think it's just not gated in this WIP form of the PR

File: src/test/ui/consts/const-eval/format_args.rs (hunk 1 above)
Comment (bjorn3) [reply src/test/ui/consts/const-eval/format_args.rs:8]: > "temporary value dropped while borrowed" is a bit odd though...

It expands to

```rust
::core::fmt::Arguments::new_v1(
    &[""],
    &match (&123,) {
        (arg0,) => [::core::fmt::ArgumentV1::new(
            arg0,
            ::core::fmt::Display::fmt,
        )],
    },
);
```

The `[::core::fmt::ArgumentV1::new(arg0, ::core::fmt::Display::fmt)]` is here borrowed and then dropped before the end of the borrow.

File: src/test/ui/consts/const-eval/format_args.rs (hunk 2 above)
Comment (m-ou-se) [reply src/test/ui/consts/const-eval/format_args.rs:3]: > So this is insta-stable?

Yes. That's why I added `needs-fcp`.

**Comment by oli-obk** [comment]
i don't think we should make this pattern special. we could just allow format_args in general, just like we allow range syntax, but there's no use for the value you get, unless you use it to panic.

File: library/core/src/fmt/mod.rs [hunk 3]
```diff
@@ -332,6 +332,15 @@ enum FlagV1 {
 }
 
 impl<'a> Arguments<'a> {
+    /// When using the format_args!() macro, this function is used to generate the
+    /// Arguments structure in case it's just a single string literal.
+    #[doc(hidden)]
+    #[inline]
+    #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
```
Comment (bjorn3) [inline library/core/src/fmt/mod.rs:339]: Is it possible to make it const-unstable at the same time under a different feature gate?

File: src/test/ui/consts/const-eval/format_args.rs (hunk 1 above)
Comment (RalfJung) [reply src/test/ui/consts/const-eval/format_args.rs:8]: ah right that would not be promoted.
Better change the test do
```
const B: () = {
  let _ = format_args!(...);
};
```

File: library/core/src/fmt/mod.rs (hunk 3 above)
Comment (RalfJung) [reply library/core/src/fmt/mod.rs:339]: It is const-unstable (implicitly). But the `format_args!` macro expanded code has some kind of span that still lets stable code call this.

Comment (bjorn3) [reply library/core/src/fmt/mod.rs:339]: Only the `fmt_internals` feature is allowed:

https://github.com/rust-lang/rust/blob/a84d1b21aea9863f0fc5f436b4982d145dade646/library/core/src/macros/mod.rs#L826

Using a different feature for const checking would require this feature to be enabled by the user of `format_args!`.

Comment (m-ou-se) [reply library/core/src/fmt/mod.rs:339]: The reason i want this is so `panic!()` can use this for const panic. (See https://github.com/rust-lang/rust/pull/86998) Putting an `allow_internal_unstable(const_fmt_internals)` on the `panic` macro won't work through the nested format_args call.

Comment (bjorn3) [reply library/core/src/fmt/mod.rs:339]: That would at least prevent it from being insta-stable, right? Maybe use `const_panic` as feature gate?

Comment (m-ou-se) [reply library/core/src/fmt/mod.rs:339]: I think we should just go for the `const_format_args!()` macro for now (#86998) and at some point stabilize `format_args!()` in const entirely. (And then remove `const_format_args!()` again.)

**Comment by Mark-Simulacrum** [comment]
@m-ou-se it seems like we can likely close this now -- I think the broader goal still probably makes sense (making format_args! const in general) and we worked around this for the panic case already, so there's not really that much need for the targeted adjustment here?

**Comment by Dylan-DPC** [comment]
Closing this based on [comment](https://github.com/rust-lang/rust/pull/87005#issuecomment-975022802)

**Pull request closed without merging**